src/cpp/session/modules/SessionRAddins.cpp

```diff
@@ -0,0 +1,315 @@
+   {
+      json::Object object;
+      
+      BOOST_FOREACH(const std::string& key, addins_ | boost::adaptors::map_keys)
+      {
+         object[key] = addins_[key].toJson();
+      }
+      
+      return object;
+   }
+
+   std::size_t size() const { return addins_.size(); }
+   
+private:
+   
+   static std::string constructKey(const std::string& package, const std::string& name)
+   {
+      return package + "::" + name;
+   }
+
+   std::map<std::string, AddinSpecification> addins_;
+};
+
+AddinRegistry& addinRegistry()
+{
+   static AddinRegistry instance;
+   return instance;
+}
+
+// TODO: This probably belongs in a separate module and could be exported
+std::vector<FilePath> getLibPaths()
+{
+   std::vector<std::string> libPathsString;
+   r::exec::RFunction rfLibPaths(".libPaths");
+   Error error = rfLibPaths.call(&libPathsString);
+   if (error)
+      LOG_ERROR(error);
+   
+   std::vector<FilePath> libPaths(libPathsString.size());
+   BOOST_FOREACH(const std::string& path, libPathsString)
+   {
+      libPaths.push_back(module_context::resolveAliasedPath(path));
+   }
+   
+   return libPaths;
+}
+
+std::map<std::string, std::string> parseAddinDcf(const std::string& contents)
+{
+   // read and parse the DCF file
+   std::map<std::string, std::string> fields;
+   std::string errMsg;
+   Error error = text::parseDcfFile(contents, true, &fields, &errMsg);
+   if (error)
+      LOG_ERROR(error);
+   
+   return fields;
+}
+
+void registerAddin(const std::string& pkgName,
+                   std::map<std::string, std::string>& fields)
+{
+   addinRegistry().add(pkgName, AddinSpecification(
+         fields["Name"],
+         pkgName,
+         fields["Title"],
+         fields["Description"],
+         fields["Binding"]));
+}
+
+void registerAddins(const std::string& pkgName, const FilePath& addinPath)
+{
+   static const boost::regex reSeparator("\\n{2,}");
+   
+   std::string contents;
+   Error error = core::readStringFromFile(addinPath, &contents, string_utils::LineEndingPosix);
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   boost::sregex_token_iterator it(contents.begin(), contents.end(), reSeparator, -1);
+   boost::sregex_token_iterator end;
+   
+   for (; it != end; ++it)
+   {
+      std::map<std::string, std::string> fields = parseAddinDcf(*it);
+      registerAddin(pkgName, fields);
+   }
+}
+
+void discoverAddins(const std::vector<FilePath>& libPaths)
+{
+   BOOST_FOREACH(const FilePath& libPath, libPaths)
+   {
+      if (!libPath.exists())
+         continue;
+      
+      // get the children in this directory
+      std::vector<FilePath> pkgPaths;
+      Error error = libPath.children(&pkgPaths);
+      if (error)
+      {
+         LOG_ERROR(error);
+         continue;
+      }
+      
+      BOOST_FOREACH(const FilePath& pkgPath, pkgPaths)
+      {
+         // construct the addin path and check if it exists
+         FilePath addinPath = pkgPath.complete("rstudio/addins.dcf");
```

Comment:
You should use childPath here -- the difference between childPath and complete is that complete will also accept an absolute path (which then is returned as the result) and complete will actually fail with an exception if the parent path doesn't exist!). In this case complete could fail if the package path was deleted after the original enumeration.